_posts/2022-04-14-javascript-immutability.md

```diff
@@ -0,0 +1,452 @@
+
+## Mutability
+
+In contrast to immutable primitive types, objects and arrays in JavaScript are **mutable**. Modifications truly modify the original data _inside_ them:
+
+```js
+let oldPerson = {name: 'Ismo'};
+let newPerson = oldPerson; // I assume I'm creating a copy... right?
+newPerson.name = 'Seppo';
+console.log(oldPerson.name); // Oops, both persons now have 'Seppo' as their name.
+```
+
+The same thing happens with arrays:
+
+```js
+let oldArray = [];
+let newArray = oldArray; // Copy or not?
+newArray.push('Hello world');
+console.log(oldArray); // Oops, both arrays now have the string 'Hello world'.
+```
+
+Objects and arrays in JavaScript are so called _reference types_. We cannot make copies of them as easily as with primitive types, we simple create a **new reference** to the original data. If we make modifications to our "copies", we actually mutate both values.
+
+_So... why is this a problem?_
+
+I don't think mutability is a problem in itself, but it can _cause_ problems if you are not careful. A typical problem is that when a data is mutable, modifications are not easy to predict and track down. This is especially true in large applications.
+
+Consider a simple example in which you have your application `state` stored as a mutable object. You and your teammates have agreed that `setNumbers` is the only way you are allowed to change the numbers in the `state`. This should make changes to the `state` predictable:
+
+```js
+let state = {numbers: [1, 2, 3], names: ['Ismo', 'Seppo']};
+let setNumbers = (newNumbers) => {
+  console.log('I\'m changing numbers to: ', newNumbers);
+  // Replace numbers in the state by creating a new state with the new numbers
+  state = {...state, numbers: newNumbers}; // We will cover this syntax later in the article
+};
+```
+
+Later, you mean to read the state and modify it as a copy:
+
+```js
+let temporaryNumbers = state.numbers;
+let temporaryNumbersPlusOne = numbers.push(1);
+// Oops, you have mutated the application state outside of setNumbers!
+```
+
+See, it is very easy to _accidentally_ modify mutable state. If the `state` object had been immutable, we could not have made such a mistake. Thus, I believe having immutable values and making modifications only by creating new versions of the old data helps us to write code that is more predictable, safer and also easier to test.
+
+_I sure did a mistake! I want to learn how to work with objects and arrays immutably!_
+
+## Immutability Using Vanilla JavaScript
+
+If a single human being has faced a problem in JavaScript and solved it, they have probably packed the solution into a JavaScript library. Things are no different with the concept of immutability: there are many possible libraries that can help you with that. The good news is that you don't necessarily need anyone of those to encourage immutability. Immutable code can be written in vanilla JavaScript. Let's take a look! 
```

Comment:
Kongruenssi:  single human being has ... he or she has ...

Tai jos haluaa välttää he/she niin jotenkin näin: "Every single JavaScript solution has probably been packaged as a library."

Anyone -> any   (anyone viittaa aina henkilöihin).